### ISSUE TITLE
Batched inplace mm changes stride when out size is correct

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#48281]
## 🐛 Bug

When `torch.matmul(A, B, out=A)` is called for batched `A` and square `B`, `A`'s strides are changed, even though it has the correct size!

## To Reproduce

![image](https://user-images.githubusercontent.com/5674597/99738873-808cb500-2a99-11eb-9207-052b69520494.png)

Code:
```py
base = torch.randn(2, 4)
x = base[..., ::2]
m = torch.zeros(2, 2)
print(f"before matmul")
print(f"base: shape={tuple(base.shape)} stride={tuple(base.stride())})")
print(f"x: shape={tuple(x.shape)} stride={tuple(x.stride())})")
print(base)
torch.matmul(x, m, out=x)
print(f"after matmul")
print(f"base: shape={tuple(base.shape)} stride={tuple(base.stride())})")
print(f"x: shape={tuple(x.shape)} stride={tuple(x.stride())})")
print(base)
```
```py
base = torch.randn(2, 2, 4)
x = base[..., ::2]
m = torch.zeros(2, 2)
print(f"before matmul")
print(f"base: shape={tuple(base.shape)} stride={tuple(base.stride())})")
print(f"x: shape={tuple(x.shape)} stride={tuple(x.stride())})")
print(base)
torch.matmul(x, m, out=x)
print(f"after matmul")
print(f"base: shape={tuple(base.shape)} stride={tuple(base.stride())})")
print(f"x: shape={tuple(x.shape)} stride={tuple(x.stride())})")
print(base)
```

**Comment by SsnL** [comment]
Note that in the given example, the innermost dimension of `A` is not contiguous. However the problem persists if it is contiguous.

**Comment by mrshenli** [comment]
@ezyang @gchanan @mruberry we used to have a `module:operator` label, but looks like it no longer exist. Any suggestions on should we label operator-related issues?

**Comment by gchanan** [comment]
`linear algebra` seems okay.

**Comment by gchanan** [comment]
it's a problem at least as far back as 1.4.0.  Looks like we need to check memory overlap.

**Comment by SsnL** [comment]
@gchanan just want to make sure, this usecase should be supported, right?

**Comment by gchanan** [comment]
yes, I was just checking if we needed to emergency fix for 1.7.1.
